cohortextractor/query_engines/base_sql.py

```diff
@@ -247,7 +245,10 @@ def get_node_list(self, node):
     def engine(self):
         if self._engine is None:
             engine_url = sqlalchemy.engine.make_url(self.backend.database_url)
-            engine_url = set_driver(engine_url)
+            # Hardcode the specific SQLAlchemy dialect we want to use: this is the
+            # dialect the query engine will have been written for and tested with and we
+            # don't want to allow global config changes to alter this
+            engine_url._get_entrypoint = lambda: self.sqlalchemy_dialect
```

Comment:
I'm a bit uncomfortable with this since it's so reliant on SQLAlchemy implementation details. But the global registry plus magic string parsing are really not helping us here.

So, what if we go with the flow? Could we make up our own driver names _per backend_? 1) register the correct driver for each backend name (even if that means registering the same driver twice with different names) and 2) insert the backend name into the URLs?

Reply:
```py
registry.register("databricks", "cohortextractor.query_engines.spark_lib", "SparkDialect")
```

and

```py
engine_url.set(drivername="databricks")
```

Reply:
Yeah, I did think about that approach. We'd have to do something like this though:
```py
registry.register(
  "<unique-name-from-somewhere>",
  self.sqlalchemy_dialect.__module__,
  self.sqlalchemy_dialect.__name__,
)
```
And that didn't really feel like an improvement to me.

I wonder if the best thing to do here is to add an assert that we really have constructed an engine with the right dialect:
```py
assert isinstance(engine.dialect, self.sqlalchemy_dialect)
```
Then if SQLAlchemy change their internals here we'll know about it straight away.

Reply:
Right. The insistence on separating the module and attribute name is unhelpful, isn't it? (And using strings not objects.)

I think your suggestion of adding an assert is a good one.